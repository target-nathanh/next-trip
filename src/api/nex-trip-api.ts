import { Direction, NexTripResult, Place, Route, Vehicle } from '../types';

const API_ROOT = 'https://svc.metrotransit.org/nextripv2';
async function getRoutes(): Promise<Route[]> {
  return baseFetch<Route[]>(`${API_ROOT}/routes`);
}

async function getDirectionsForRoute(routeId: string): Promise<Direction[]> {
  return baseFetch<Direction[]>(`${API_ROOT}/directions/${routeId}`);
}

async function getPlaces(routeId: string, directionId: number): Promise<Place[]> {
  return baseFetch<Place[]>(`${API_ROOT}/stops/${routeId}/${directionId}`);
}

async function getNexTripResult(
  routeId: string,
  directionId: number,
  placeCode: string
): Promise<NexTripResult> {
  return baseFetch<NexTripResult>(`${API_ROOT}/${routeId}/${directionId}/${placeCode}`);
}

async function getNexTripResultByStopId(stopId: string): Promise<NexTripResult> {
  return baseFetch<NexTripResult>(`${API_ROOT}/${stopId}`);
}

async function getVehiclesByRoute(routeId: string): Promise<Vehicle[]> {
  return baseFetch<Vehicle[]>(`${API_ROOT}/vehicles/${routeId}`);
}

async function getDirectionsAndPlaces(
  routeId: string,
  directionId: number
): Promise<{ places: Place[]; directions: Direction[] }> {
  const directions = await getDirectionsForRoute(routeId);
  const places = await getPlaces(routeId, directionId);

  return { places, directions };
}

async function baseFetch<T>(url: string, body?: any, extraHeaders = {}): Promise<T> {
  const fetchConfig: { body?: string; headers: Record<string, string> } = {
    headers: {
      'Content-Type': 'application/json',
      ...extraHeaders,
    },
  };
  if (body) {
    fetchConfig.body = JSON.stringify(body);
  }
  const response = await fetch(url, fetchConfig);

  if (response.ok) {
    return await response.json();
  }

  const error = new Error(response.status.toString());
  error.message = response.statusText;
  throw error;
}

export const NexTripApi = {
  getRoutes,
  getDirectionsForRoute,
  getNexTripResult,
  getPlaces,
  getNexTripResultByStopId,
  getVehiclesByRoute,
  getDirectionsAndPlaces,
};
