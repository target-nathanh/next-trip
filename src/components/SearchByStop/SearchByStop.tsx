import TextField from '@material-ui/core/TextField';
import Icon from '@material-ui/core/Icon';
import React, { KeyboardEvent, useState } from 'react';
import { Container, IconButton, makeStyles } from '@material-ui/core';
import { NexTripApi } from '../../api/nex-trip-api';
import { Route, Switch, useHistory, useRouteMatch } from 'react-router-dom';
import Results from '../Results/Results';

const useStyles = makeStyles({
  input: {
    '& input::-webkit-outer-spin-button, & input::-webkit-inner-spin-button': {
      display: 'none',
      margin: 80,
    },
  },
  inputContainer: {
    marginTop: '10px',
    marginBottom: '10px',
  },
});

const SearchByStop: React.FC = () => {
  const { path, url } = useRouteMatch();
  const history = useHistory();
  const { input, inputContainer } = useStyles();
  const [stopNumber, setStopNumber] = useState<string>('');
  const [showError, setShowError] = useState<boolean>(false);

  const onSearch = async () => {
    try {
      const results = await NexTripApi.getNexTripResultByStopId(stopNumber);
      history.push({ pathname: `${url}/${stopNumber}`, state: { results } });
    } catch (e) {
      setShowError(true);
    }
  };
  const onKeyDownHandler = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'Enter') {
      onSearch();
    }
  };

  return (
    <Container className={inputContainer}>
      <TextField
        variant="outlined"
        className={input}
        error={showError}
        aria-errormessage={'The entered stop number is not valid'}
        type="number"
        value={stopNumber}
        onChange={(event) => setStopNumber(event.target.value)}
        onKeyDown={(event) => onKeyDownHandler(event)}
        placeholder="Enter stop number"
      />
      <IconButton data-testid="search-icon" aria-label="search" onClick={() => onSearch()}>
        <Icon>search</Icon>
      </IconButton>

      <Switch>
        <Route path={`${path}/:stopId`}>
          <Results />
        </Route>
      </Switch>
    </Container>
  );
};

export default SearchByStop;
