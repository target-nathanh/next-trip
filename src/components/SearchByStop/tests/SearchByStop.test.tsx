import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import SearchByStop from '../SearchByStop';
import { MemoryRouter } from 'react-router-dom';
import { NexTripApi } from '../../../api/nexTripApi';

describe('SearchByStop', () => {
  test('matches snapshot', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/stop']}>
        <SearchByStop />
      </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls the api with the selected stop id on clicking the icon', () => {
    NexTripApi.getNexTripResultByStopId = jest.fn().mockResolvedValue({});

    render(
      <MemoryRouter initialEntries={['/stop']}>
        <SearchByStop />
      </MemoryRouter>
    );
    const input = screen.getByPlaceholderText('Enter stop number');
    const icon = screen.getByTestId('search-icon');

    fireEvent.change(input, { target: { value: '16137' } });
    fireEvent.click(icon);

    expect(NexTripApi.getNexTripResultByStopId).toHaveBeenCalledWith('16137');
  });

  test('calls the api with the selected stop id on hitting enter', () => {
    NexTripApi.getNexTripResultByStopId = jest.fn().mockResolvedValue({});

    render(
      <MemoryRouter initialEntries={['/stop']}>
        <SearchByStop />
      </MemoryRouter>
    );
    const input = screen.getByPlaceholderText('Enter stop number');

    fireEvent.change(input, { target: { value: '16137' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(NexTripApi.getNexTripResultByStopId).toHaveBeenCalledWith('16137');
  });
});
