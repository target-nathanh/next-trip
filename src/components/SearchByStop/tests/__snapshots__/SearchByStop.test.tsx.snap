// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchByStop matches snapshot 1`] = `
<div
  class="MuiContainer-root makeStyles-inputContainer-2 MuiContainer-maxWidthLg"
>
  <div
    aria-errormessage="The entered stop number is not valid"
    class="MuiFormControl-root MuiTextField-root makeStyles-input-1"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        placeholder="Enter stop number"
        type="number"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-4"
          style="width: 0.01px;"
        >
          <span>
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    aria-label="search"
    class="MuiButtonBase-root MuiIconButton-root"
    data-testid="search-icon"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiIconButton-label"
    >
      <span
        aria-hidden="true"
        class="material-icons MuiIcon-root"
      >
        search
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
